pkg/ingester/user_tsdb.go

```diff
@@ -237,19 +239,35 @@ func (u *userTSDB) blocksToDelete(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
 	if u.db == nil {
 		return nil
 	}
+
 	deletable := tsdb.DefaultBlocksToDelete(u.db)(blocks)
-	if u.shipper == nil {
-		return deletable
+	result := map[ulid.ULID]struct{}{}
+	deadline := time.Now().Add(-u.blockMinRetention)
+
+	// The shipper enabled case goes first because its common in the way we run the ingesters
+	if u.shipper != nil {
+		shippedBlocks := u.getCachedShippedBlocks()
+
+		for blockID, blockShippedTime := range shippedBlocks {
+			if blockShippedTime.Before(deadline) {
+				if _, ok := deletable[blockID]; ok {
```

Comment:
[nit] Now that `shippedBlocks` is a map, we could iterate over `deletable` and check if it was shipped. The logic would be clearer that way, but it's not a blocker.